runtime/compiler/runtime/J9CodeCacheManager.cpp

```diff
@@ -355,18 +359,39 @@ J9::CodeCacheManager::allocateCodeCacheSegment(size_t segmentSize,
 
    void *defaultEndAddress = vmemParams.endAddress;
 
+   uintptr_t someJitLibraryAddress = self()->getSomeJitLibraryAddress();
+
+#if defined(TR_TARGET_POWER)
+   size_t alignment = 64 * 1024;
+#elif (defined(LINUX) && defined(TR_TARGET_X86))
+   size_t alignment = 2 * 1024 * 1024;
+#elif (defined(TR_TARGET_S390))
+   size_t alignment = 1024 * 1024;
+#else
+   size_t alignment = pageSizes[0];
+#endif
+   if (largeCodePageSize > 0)
+      alignment = largeCodePageSize;
+
    if (preferredStartAddress)
       {
-   // J9PORT_VMEM_ADDRESS_HINT is only supported for default pages and linux
+      vmemParams.options |= J9PORT_VMEM_STRICT_ADDRESS; // if we cannot allocate a block in the desired range, return NULL
+      vmemParams.alignmentInBytes = alignment;
 #if defined(LINUX)
-      if (largeCodePageSize == 0)
+      vmemParams.options |= J9PORT_VMEM_ALLOC_QUICK; // Use smaps for guidance
+      vmemParams.startAddress = preferredStartAddress;
+      if ((uintptr_t)preferredStartAddress < someJitLibraryAddress) // Allocate before JIT dll
          {
-         vmemParams.options |= J9PORT_VMEM_ADDRESS_HINT;
+         vmemParams.endAddress = (void *)(someJitLibraryAddress - SAFE_DISTANCE_REPOSITORY_JITLIBRARY - segmentSize);
          }
+      else // Allocate after JIT dll
+         {
+         vmemParams.endAddress = (void *)(someJitLibraryAddress + MAX_DISTANCE_NEAR_JITLIBRARY_TO_AVOID_TRAMPOLINE - segmentSize);
+         }
```

Comment:
Since this code only runs once, maybe it's worth putting a `TR_ASSERT_FATAL` to ensure that `endAddress` is always greater than `startAddress`?